recipes/newrelic/infrastructure/ohi/postgres/debian.yml

```diff
@@ -95,101 +105,67 @@ install:
+          BEGIN
+            IF NOT EXISTS (SELECT FROM pg_catalog.pg_roles WHERE  rolname = 'newrelic') THEN
+              CREATE ROLE newrelic LOGIN PASSWORD '$GEN_PASSWORD';
+            END IF;
+            ALTER USER newrelic WITH PASSWORD '$GEN_PASSWORD';
+            GRANT SELECT ON pg_stat_database TO \"$NR_CLI_DB_USERNAME\";
+            GRANT SELECT ON pg_stat_database_conflicts TO \"$NR_CLI_DB_USERNAME\";
+            GRANT SELECT ON pg_stat_bgwriter TO \"$NR_CLI_DB_USERNAME\";
+          END
+          \$do\$;
 
-              IS_DATABASE_VALID=$(PGPASSWORD=$NR_CLI_DB_PASSWORD psql -U $NR_CLI_DB_USERNAME -w $NR_CLI_DATABASE -h $NR_CLI_DB_HOSTNAME -p $NR_CLI_DB_PORT -c "select version();" 2>&1 | grep -i PostgreSQL | grep -v grep | wc -l)
-              if [ $IS_DATABASE_VALID -eq 0 ] ; then
-                printf "\n[Error]: The provided database name is not accessible with the provided username and password and port.\nSee https://docs.newrelic.com/docs/integrations/host-integrations/host-integrations-list/postgresql-monitoring-integration#comp-req for more info.\n" >&2
-                if [ ! $TRIES -lt {{.MAX_RETRIES}} ]; then exit 2; else continue; fi
-                echo "Please try again"
-              fi
-              break
-            done
-            printf "\nConnectivity and User Credentials are valid!\n\n"
-          fi
+          EOT
+          )
+              sudo -u postgres psql -t -A -c "${sql}" > /dev/null 2>&1 || true
+              sudo rm -f .nr_script.sql > /dev/null 2>&1
+            fi
+            IS_DATABASE_VALID=$(PGPASSWORD=$NR_CLI_DB_PASSWORD psql -U $NR_CLI_DB_USERNAME -w $NR_CLI_DATABASE -h $NR_CLI_DB_HOSTNAME -p $NR_CLI_DB_PORT -c "select version();" 2>&1 | grep -i PostgreSQL | grep -v grep | wc -l)
+            IS_DATABASE_VALID=${IS_DATABASE_VALID:-0}
+          } 
 
+          try_create_user
           TRIES=0
-          if [[ "$NEW_RELIC_ASSUME_YES" != "true" ]]; then
-            while [ $TRIES -lt {{.MAX_RETRIES}} ]; do
-              printf "Use SSL to connect? (false/true) (default: false): "
-              read -r NR_CLI_SSL
-              NR_CLI_SSL=${NR_CLI_SSL:-false}
-              ((TRIES++))
-              if [[ "$NR_CLI_SSL" == "true" ]]; then
-                printf "Client Certificate file (if applicable): "
-                read -r NR_CLI_CLIENT_CERT_FILE
-                printf "PEM key file (if applicable): "
-                read -r NR_CLI_CERT_KEY
-                printf "Trust certificate? (false/true) (default: false): "
-                read -r NR_CLI_TRUST_SERVER_CERTIFICATE
-                NR_CLI_TRUST_SERVER_CERTIFICATE=${NR_CLI_TRUST_SERVER_CERTIFICATE:-false}
-                if [[ "$NR_CLI_TRUST_SERVER_CERTIFICATE" == "true" ]]; then
-                  printf "Certificate Authority file (if applicable): "
-                  read -r NR_CLI_CERT_AUTH_FILE
-                fi
+          while [[ $TRIES -lt {{.MAX_RETRIES}} && $IS_DATABASE_VALID -eq "0" ]]; do
```

Comment:
I'm not sure why we're looping here, since it will redo the exact same call to try_create_user.
Unless we're prompting for a username/password the user may have created, and we want to check we can connect, I don't think we need to loop, and perhaps we don't need a try_create_user function and could just process the creation of the user inline?

Reply:
@Julien4218  The loop is to re-prompt the user for postgres user password.  If they miss typing the first type when it tries to create a user, we give them a few more chances to type it in correctly.

Reply:
@NRhzhao ho I understand now, thanks.
It might be good to add a comment to explain, I don't know that it's obvious enough for when we'll look at it down the road.
Also if you could split the part that fetches IS_DATABASE_VALID from the try_create_user function, it would be a bit more readable. It's not obvious at first glance the function has the side effect of setting the variable. You may even consider removing the function and just put it inline in the loop, since it's only called from within the loop.